#### **Introduction**

The objective of this project was to investigate potential correlations between specific player attributes of 2023 NFL teams—such as average age and weight across different positions—and key performance metrics from the 2023 season, including points scored (PF), points allowed (PA), and yards per return (Y/Rt). Specifically, the analysis aimed to explore the following relationships:

Offensive Attributes: Correlation between the average age and weight of offensive players and the total points scored by the team (PF).
Defensive Attributes: Correlation between the average age and weight of defensive players and the total points allowed by the team (PA).
Special Teams Attributes: Correlation between the average age and weight of special teams players and the average yards per return (Y/Rt).

By evaluating these relationships, I aim  determine if there were any statistically significant correlations that might warrant further investigation. Such insights can be valuable for betting strategies and sports analysis, especially since these relationships might exist only for short periods, such as within a single season or two. The project begins with the assumption that it is highly unlikely that these meaninful correlations will be evident and thus a small but significant samople size is first tested. 

The analysis focused on four AFC East NFL teams: the Buffalo Bills, Miami Dolphins, New York Jets, and New England Patriots. Data was obtained from Pro Football Reference, converted into CSV files, and imported into a Python environment for analysis. Python's pandas and numpy libraries were used extensively for data manipulation, calculation of averages, normalization of data, and application of statistical tests.


CSV files containing player data for each team were loaded into Python using pandas. The datasets were processed to use the correct headers and remove any extraneous rows.
The player data was grouped by position, and calculations were made for the average age and weight for each position group (e.g., offensive, defensive, special teams).
Categorization of Positions:

Positions were categorized into offense, defense, and special teams to facilitate focused analysis. Each position was mapped to a consolidated category (e.g., quarterbacks, running backs, defensive backs) to ensure a consistent and meaningful comparison across teams.
Normalization:

The average age and weight for each position within each team were normalized to account for potential differences in scale and distribution. This allowed for a fair comparison of metrics across teams and positions.
Correlation Analysis:

The Pearson correlation coefficient was applied to evaluate the linear relationship between player attributes (average age and weight) and team performance metrics (PF, PA, Y/Rt).
The correlation coefficient (r) indicates the strength and direction of the relationship, while the p-value assesses the statistical significance of the observed correlation.
Understanding Correlation and Statistical Significance
Correlation Coefficient (r):

Measures the strength and direction of a linear relationship between two variables. A value close to 1 or -1 indicates a strong positive or negative correlation, respectively, while a value close to 0 suggests little or no linear relationship.
P-Value:

The p-value helps determine the significance of the observed correlation. It represents the probability that the observed correlation could have occurred by random chance, assuming no actual relationship exists in the population.
A low p-value (typically less than 0.05) suggests that the observed correlation is unlikely to have occurred by chance, indicating a statistically significant relationship.
Why the P-Value Matters
Even when the correlation coefficient (r) appears moderately strong (e.g., 0.59), a high p-value (e.g., 0.41) suggests there is a substantial probability (41%) that the observed correlation could have occurred by random chance. This means we cannot confidently assert that the observed correlation reflects a true relationship in the population; it could simply result from random variation in the sample data.

To draw meaningful conclusions, both the correlation coefficient and the p-value should be considered together:

A high correlation with a low p-value indicates a statistically significant and potentially meaningful relationship.
A high correlation with a high p-value suggests that, although there seems to be a relationship, the evidence is insufficient to rule out random chance.
Conversely, a low correlation with a low p-value still indicates a weak but statistically significant relationship, which might not be practically meaningful.

In [518]:
import pandas as pd

# Load the CSV file, using the second row as the header
df = pd.read_csv('bills.csv', header=1)

# Print the first few rows of the data
print(df.head())

    No.             Player   Age Pos   G    GS     Wt    Ht      College/Univ  \
0  17.0         Josh Allen  27.0  QB  17  17.0  237.0   6-5    ReedleyWyoming   
1   9.0         Kyle Allen  27.0  QB   7   0.0  210.0   6-3  Texas A&MHouston   
2   2.0         Tyler Bass  26.0   K  17   0.0  183.0  5-10  Georgia Southern   
3  71.0         Ryan Bates  26.0   G  17   0.0  302.0   6-4          Penn St.   
4  47.0  Christian Benford  23.0  CB  15  14.0  205.0   6-1         Villanova   

   BirthDate Yrs    AV                      Drafted (tm/rnd/yr)  \
0  5/21/1996   5  18.0    Buffalo Bills / 1st / 7th pick / 2018   
1   3/8/1996   5   0.0                                      NaN   
2  2/14/1997   3   3.0  Buffalo Bills / 6th / 188th pick / 2020   
3  2/14/1997   4   1.0                                      NaN   
4  9/21/2000   1   5.0  Buffalo Bills / 6th / 185th pick / 2022   

  Player-additional  
0          AlleJo02  
1          AlleKy00  
2          BassTy00  
3          BateRy02  


In [519]:
import pandas as pd

# Load the CSV files
files = ['bills.csv', 'dolphins.csv', 'patriots.csv', 'jets.csv']
dfs = [pd.read_csv(file, header=1) for file in files]

# Concatenate the DataFrames
df = pd.concat(dfs)

# Get unique positions
unique_positions = df['Pos'].unique()

# Print the unique positions
print(unique_positions)

['QB' 'K' 'G' 'CB' 'LB' 'T' 'RB' 'WR' 'DE' 'LS' 'RDE' 'DT' 'FB' 'S' 'TE'
 'P' 'C' 'SS' nan 'LILB' 'DB' 'OL' 'LCB' 'RG' 'NT' 'LOLB' 'LDE' 'DL' 'RCB']


In [520]:
import pandas as pd

# Load the bills.csv file, using the second row as the header
df_bills = pd.read_csv('bills.csv', header=1)

# Calculate the average Age and Weight for each position
avg_stats_by_position = df_bills.groupby('Pos').agg(
    Average_Age=('Age', 'mean'),
    Average_Weight=('Wt', 'mean')
).reset_index()

# Display the results
print(avg_stats_by_position)


    Pos  Average_Age  Average_Weight
0     C    31.000000      305.000000
1    CB    27.200000      194.700000
2    DE    26.000000      269.200000
3    DT    29.666667      314.833333
4    FB    26.000000      244.000000
5     G    25.250000      318.750000
6     K    26.000000      183.000000
7    LB    27.750000      234.000000
8    LS    29.000000      235.000000
9     P    33.000000      211.000000
10   QB    27.000000      223.500000
11   RB    27.400000      214.200000
12  RDE    31.000000      240.000000
13    S    28.000000      201.666667
14   SS    32.000000      191.000000
15    T    26.333333      312.666667
16   TE    25.000000      248.666667
17   WR    26.166667      195.000000


In [521]:
import numpy as np

# Compute the mean and standard deviation for the 'Average_Age' and 'Average_Weight' columns
mean_avg_age = np.mean(avg_stats_by_position['Average_Age'])
std_avg_age = np.std(avg_stats_by_position['Average_Age'])

mean_avg_weight = np.mean(avg_stats_by_position['Average_Weight'])
std_avg_weight = np.std(avg_stats_by_position['Average_Weight'])

# Normalize the 'Average_Age' and 'Average_Weight' columns using vectorized operations
avg_stats_by_position['Normalized_Average_Age'] = (avg_stats_by_position['Average_Age'] - mean_avg_age) / std_avg_age
avg_stats_by_position['Normalized_Average_Weight'] = (avg_stats_by_position['Average_Weight'] - mean_avg_weight) / std_avg_weight

# Display the normalized values
print(avg_stats_by_position[['Pos', 'Average_Age', 'Normalized_Average_Age', 'Average_Weight', 'Normalized_Average_Weight']])


    Pos  Average_Age  Normalized_Average_Age  Average_Weight  \
0     C    31.000000                1.281251      305.000000   
1    CB    27.200000               -0.334684      194.700000   
2    DE    26.000000               -0.844980      269.200000   
3    DT    29.666667                0.714256      314.833333   
4    FB    26.000000               -0.844980      244.000000   
5     G    25.250000               -1.163914      318.750000   
6     K    26.000000               -0.844980      183.000000   
7    LB    27.750000               -0.100799      234.000000   
8    LS    29.000000                0.430759      235.000000   
9     P    33.000000                2.131743      211.000000   
10   QB    27.000000               -0.419734      223.500000   
11   RB    27.400000               -0.249635      214.200000   
12  RDE    31.000000                1.281251      240.000000   
13    S    28.000000                0.005512      201.666667   
14   SS    32.000000                1.70

In [522]:
import pandas as pd
import numpy as np

# List of team files
files = ['bills.csv', 'dolphins.csv', 'patriots.csv', 'jets.csv']

# Function to calculate and normalize metrics
def process_team(file):
    # Load the dataset
    df = pd.read_csv(file, header=1)

    # Calculate the average Age and Weight for each position
    avg_stats = df.groupby('Pos').agg(
        Average_Age=('Age', 'mean'),
        Average_Weight=('Wt', 'mean')
    ).reset_index()

    # Compute mean and standard deviation for normalization
    mean_avg_age = np.mean(avg_stats['Average_Age'])
    std_avg_age = np.std(avg_stats['Average_Age'])
    mean_avg_weight = np.mean(avg_stats['Average_Weight'])
    std_avg_weight = np.std(avg_stats['Average_Weight'])

    # Normalize the averages
    avg_stats['Normalized_Average_Age'] = (avg_stats['Average_Age'] - mean_avg_age) / std_avg_age
    avg_stats['Normalized_Average_Weight'] = (avg_stats['Average_Weight'] - mean_avg_weight) / std_avg_weight

    # Return the processed DataFrame
    return avg_stats

# Process each team's data and store the results in a dictionary
team_results = {file.split('.')[0].capitalize(): process_team(file) for file in files}

# Display results for each team
for team, data in team_results.items():
    print(f"\nAverage Age and Weight by Position for {team}:")
    print(data[['Pos', 'Average_Age', 'Normalized_Average_Age', 'Average_Weight', 'Normalized_Average_Weight']])



Average Age and Weight by Position for Bills:
    Pos  Average_Age  Normalized_Average_Age  Average_Weight  \
0     C    31.000000                1.281251      305.000000   
1    CB    27.200000               -0.334684      194.700000   
2    DE    26.000000               -0.844980      269.200000   
3    DT    29.666667                0.714256      314.833333   
4    FB    26.000000               -0.844980      244.000000   
5     G    25.250000               -1.163914      318.750000   
6     K    26.000000               -0.844980      183.000000   
7    LB    27.750000               -0.100799      234.000000   
8    LS    29.000000                0.430759      235.000000   
9     P    33.000000                2.131743      211.000000   
10   QB    27.000000               -0.419734      223.500000   
11   RB    27.400000               -0.249635      214.200000   
12  RDE    31.000000                1.281251      240.000000   
13    S    28.000000                0.005512      201.666

In [523]:
import pandas as pd

# Define position categories
position_categories = {
    'Offense': ['QB', 'RB', 'WR', 'TE', 'FB', 'OL', 'T', 'G', 'C', 'RG', 'LG'],
    'Defense': ['CB', 'S', 'SS', 'FS', 'LB', 'MLB', 'OLB', 'DE', 'DT', 'DL', 'NT', 'DB', 'LCB', 'RCB', 'LDE', 'RDE', 'LOLB', 'LILB'],
    'Special Teams': ['K', 'P', 'LS']
}

# Function to assign categories
def assign_category(pos):
    for category, positions in position_categories.items():
        if pos in positions:
            return category
    return 'Unknown'

# Function to process each team's data and assign categories
def process_team(file):
    # Load the dataset
    df = pd.read_csv(file, header=1)

    # Assign categories to each position
    df['Category'] = df['Pos'].apply(assign_category)
    
    # Return the DataFrame with the category assigned
    return df

# List of team files
files = ['bills.csv', 'dolphins.csv', 'patriots.csv', 'jets.csv']

# Process each team's data and store the results in a list
team_results = {file.split('.')[0].capitalize(): process_team(file) for file in files}

# Display results for each team
for team, data in team_results.items():
    print(f"\nData for {team} with Assigned Categories:")
    print(data[['Pos', 'Category']])  # Display all rows for 'Pos' and 'Category' columns



Data for Bills with Assigned Categories:
    Pos       Category
0    QB        Offense
1    QB        Offense
2     K  Special Teams
3     G        Offense
4    CB        Defense
..  ...            ...
58    T        Offense
59   DE        Defense
60   CB        Defense
61   LB        Defense
62  NaN        Unknown

[63 rows x 2 columns]

Data for Dolphins with Assigned Categories:
    Pos       Category
0    RB        Offense
1    RB        Offense
2    CB        Defense
3     T        Offense
4     P  Special Teams
..  ...            ...
66   WR        Offense
67   RB        Offense
68   DE        Defense
69    T        Offense
70  NaN        Unknown

[71 rows x 2 columns]

Data for Patriots with Assigned Categories:
    Pos       Category
0    OL        Offense
1     C        Offense
2     C        Offense
3    CB        Defense
4     P  Special Teams
..  ...            ...
68   LB        Defense
69   CB        Defense
70   DL        Defense
71   QB        Offense
72  NaN        Un

In [524]:
import pandas as pd

# Define position categories
position_categories = {
    'Offense': ['QB', 'RB', 'WR', 'TE', 'FB', 'OL', 'T', 'G', 'C', 'RG', 'LG'],
    'Defense': ['CB', 'S', 'SS', 'FS', 'LB', 'MLB', 'OLB', 'DE', 'DT', 'DL', 'NT', 'DB', 'LCB', 'RCB', 'LDE', 'RDE', 'LOLB', 'LILB'],
    'Special Teams': ['K', 'P', 'LS']
}

# Function to assign categories
def assign_category(pos):
    for category, positions in position_categories.items():
        if pos in positions:
            return category
    return 'Unknown'

# Function to process each team's data, assign categories, and remove unknown values
def process_team(file):
    # Load the dataset
    df = pd.read_csv(file, header=1)

    # Assign categories to each position
    df['Category'] = df['Pos'].apply(assign_category)

    # Remove rows with 'Unknown' in the 'Category' column
    df = df[df['Category'] != 'Unknown']
    
    # Return the DataFrame with the category assigned and unknowns removed
    return df

# List of team files
files = ['bills.csv', 'dolphins.csv', 'patriots.csv', 'jets.csv']

# Process each team's data and store the results in a dictionary
team_results = {file.split('.')[0].capitalize(): process_team(file) for file in files}

# Display results for each team without unknowns
for team, data in team_results.items():
    print(f"\nData for {team} with Assigned Categories (Unknowns Removed):")
    print(data[['Pos', 'Category']])



Data for Bills with Assigned Categories (Unknowns Removed):
   Pos       Category
0   QB        Offense
1   QB        Offense
2    K  Special Teams
3    G        Offense
4   CB        Defense
..  ..            ...
57   G        Offense
58   T        Offense
59  DE        Defense
60  CB        Defense
61  LB        Defense

[62 rows x 2 columns]

Data for Dolphins with Assigned Categories (Unknowns Removed):
   Pos       Category
0   RB        Offense
1   RB        Offense
2   CB        Defense
3    T        Offense
4    P  Special Teams
..  ..            ...
65  OL        Offense
66  WR        Offense
67  RB        Offense
68  DE        Defense
69   T        Offense

[70 rows x 2 columns]

Data for Patriots with Assigned Categories (Unknowns Removed):
   Pos       Category
0   OL        Offense
1    C        Offense
2    C        Offense
3   CB        Defense
4    P  Special Teams
..  ..            ...
67  DE        Defense
68  LB        Defense
69  CB        Defense
70  DL        Def

In [525]:
pip install tabulate



[notice] A new release of pip is available: 24.1.1 -> 24.2
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [526]:
import pandas as pd
from tabulate import tabulate

# Define position categories
position_categories = {
    'Offense': ['QB', 'RB', 'WR', 'TE', 'FB', 'OL', 'T', 'G', 'C', 'RG', 'LG'],
    'Defense': ['CB', 'S', 'SS', 'FS', 'LB', 'MLB', 'OLB', 'DE', 'DT', 'DL', 'NT', 'DB', 'LCB', 'RCB', 'LDE', 'RDE', 'LOLB', 'LILB'],
    'Special Teams': ['K', 'P', 'LS']
}

# Function to assign categories
def assign_category(pos):
    for category, positions in position_categories.items():
        if pos in positions:
            return category
    return 'Unknown'

# Assume 'team_results' contains pre-processed data for each team

# Add 'Team' column to each team's DataFrame in 'team_results'
for team_name, df in team_results.items():
    df['Team'] = team_name  # Assign the team name to each DataFrame

# Combine all processed data into a single DataFrame
all_data = pd.concat(team_results.values(), ignore_index=True)

# Assign categories for clarity
all_data['Category'] = all_data['Pos'].apply(assign_category)

# Properly aggregate data to get average Age and Weight for each position within each team
position_averages = all_data.groupby(['Category', 'Pos', 'Team']).agg(
    Average_Age=('Age', 'mean'),
    Average_Weight=('Wt', 'mean')
).reset_index()

# Normalize the averages by position
position_averages['Normalized_Age'] = position_averages.groupby('Pos')['Average_Age'].transform(
    lambda x: (x - x.mean()) / x.std())
position_averages['Normalized_Weight'] = position_averages.groupby('Pos')['Average_Weight'].transform(
    lambda x: (x - x.mean()) / x.std())

# Rank the teams for each position by actual values (age and weight)
position_averages['Rank_Age'] = position_averages.groupby('Pos')['Average_Age'].rank(ascending=False)
position_averages['Rank_Weight'] = position_averages.groupby('Pos')['Average_Weight'].rank(ascending=False)

# Function to display the data grouped by category and position
def display_grouped_data(data):
    for category in ['Offense', 'Defense', 'Special Teams']:
        category_data = data[data['Category'] == category]
        print(f"\n=== {category} ===")
        for position in category_data['Pos'].unique():
            pos_data = category_data[category_data['Pos'] == position]
            # Convert to list of lists for tabulate
            table_data = pos_data[['Team', 'Average_Age', 'Normalized_Age', 'Rank_Age',
                                   'Average_Weight', 'Normalized_Weight', 'Rank_Weight']].values.tolist()
            headers = ['Team', 'Avg Age', 'Norm Age', 'Rank Age', 'Avg Wt', 'Norm Wt', 'Rank Wt']
            print(f"\nPosition: {position}")
            print(tabulate(table_data, headers=headers, tablefmt='grid'))

# Display the aggregated, normalized, and ranked data
display_grouped_data(position_averages)



=== Offense ===

Position: C
+----------+-----------+------------+------------+----------+-----------+-----------+
| Team     |   Avg Age |   Norm Age |   Rank Age |   Avg Wt |   Norm Wt |   Rank Wt |
+==========+===========+============+============+==========+===========+===========+
| Bills    |      31   |  0.955847  |          1 |      305 | -0.473016 |         2 |
+----------+-----------+------------+------------+----------+-----------+-----------+
| Jets     |      23   | -1.03896   |          3 |      313 |  1.14875  |         1 |
+----------+-----------+------------+------------+----------+-----------+-----------+
| Patriots |      27.5 |  0.0831172 |          2 |      304 | -0.675737 |         3 |
+----------+-----------+------------+------------+----------+-----------+-----------+

Position: FB
+----------+-----------+------------+------------+----------+-----------+-----------+
| Team     |   Avg Age |   Norm Age |   Rank Age |   Avg Wt |   Norm Wt |   Rank Wt |
+=========

In [527]:
import pandas as pd
from tabulate import tabulate

# Define consolidated position categories
consolidated_categories = {
    'Offensive Line': ['OL', 'T', 'G', 'C', 'RG', 'LG'],
    'Quarterbacks': ['QB'],
    'Running Backs': ['RB', 'FB'],
    'Tight Ends': ['TE'],
    'Receivers': ['WR'],
    'Defensive Line': ['DL', 'DE', 'DT', 'NT', 'LDE', 'RDE'],
    'Linebackers': ['LB', 'MLB', 'OLB', 'LOLB', 'LILB'],
    'Defensive Backs': ['CB', 'S', 'SS', 'FS', 'DB', 'LCB', 'RCB'],
    'Special Teams': ['K', 'P', 'LS']
}

# Function to assign consolidated categories
def assign_consolidated_category(pos):
    for category, positions in consolidated_categories.items():
        if pos in positions:
            return category
    return 'Unknown'

# Assume 'team_results' contains pre-processed data for each team

# Add 'Team' column to each team's DataFrame in 'team_results'
for team_name, df in team_results.items():
    df['Team'] = team_name  # Assign the team name to each DataFrame

# Combine all processed data into a single DataFrame
all_data = pd.concat(team_results.values(), ignore_index=True)

# Assign consolidated categories
all_data['Consolidated_Category'] = all_data['Pos'].apply(assign_consolidated_category)

# Remove 'Unknown' categories (if any)
all_data = all_data[all_data['Consolidated_Category'] != 'Unknown']

# Aggregate data to get average Age and Weight for each consolidated category within each team
consolidated_averages = all_data.groupby(['Consolidated_Category', 'Team']).agg(
    Average_Age=('Age', 'mean'),
    Average_Weight=('Wt', 'mean')
).reset_index()

# Normalize the averages by consolidated category
consolidated_averages['Normalized_Age'] = consolidated_averages.groupby('Consolidated_Category')['Average_Age'].transform(
    lambda x: (x - x.mean()) / x.std())
consolidated_averages['Normalized_Weight'] = consolidated_averages.groupby('Consolidated_Category')['Average_Weight'].transform(
    lambda x: (x - x.mean()) / x.std())

# Rank the teams for each consolidated category by actual values (age and weight)
consolidated_averages['Rank_Age'] = consolidated_averages.groupby('Consolidated_Category')['Average_Age'].rank(ascending=False)
consolidated_averages['Rank_Weight'] = consolidated_averages.groupby('Consolidated_Category')['Average_Weight'].rank(ascending=False)

# Function to display the data grouped by consolidated category
def display_consolidated_data(data):
    for category in consolidated_categories.keys():
        category_data = data[data['Consolidated_Category'] == category]
        print(f"\n=== {category} ===")
        # Convert to list of lists for tabulate
        table_data = category_data[['Team', 'Average_Age', 'Normalized_Age', 'Rank_Age',
                                   'Average_Weight', 'Normalized_Weight', 'Rank_Weight']].values.tolist()
        headers = ['Team', 'Avg Age', 'Norm Age', 'Rank Age', 'Avg Wt', 'Norm Wt', 'Rank Wt']
        print(tabulate(table_data, headers=headers, tablefmt='grid'))

# Display the aggregated, normalized, and ranked data for consolidated categories
display_consolidated_data(consolidated_averages)



=== Offensive Line ===
+----------+-----------+------------+------------+----------+-----------+-----------+
| Team     |   Avg Age |   Norm Age |   Rank Age |   Avg Wt |   Norm Wt |   Rank Wt |
+==========+===========+============+============+==========+===========+===========+
| Bills    |   26.375  |  -1.271    |          4 |  314.75  |  0.234372 |         2 |
+----------+-----------+------------+------------+----------+-----------+-----------+
| Dolphins |   27.3636 |   0.400027 |          2 |  310.818 | -0.884861 |         4 |
+----------+-----------+------------+------------+----------+-----------+-----------+
| Jets     |   27      |  -0.214604 |          3 |  311.6   | -0.662309 |         3 |
+----------+-----------+------------+------------+----------+-----------+-----------+
| Patriots |   27.7692 |   1.08558  |          1 |  318.538 |  1.3128   |         1 |
+----------+-----------+------------+------------+----------+-----------+-----------+

=== Quarterbacks ===
+-------

In [528]:
import pandas as pd
from tabulate import tabulate

# Define consolidated position categories
consolidated_categories = {
    'Offensive Line': ['OL', 'T', 'G', 'C', 'RG', 'LG'],
    'Quarterbacks': ['QB'],
    'Running Backs': ['RB', 'FB'],
    'Tight Ends': ['TE'],
    'Receivers': ['WR'],
    'Defensive Line': ['DL', 'DE', 'DT', 'NT', 'LDE', 'RDE'],
    'Linebackers': ['LB', 'MLB', 'OLB', 'LOLB', 'LILB'],
    'Defensive Backs': ['CB', 'S', 'SS', 'FS', 'DB', 'LCB', 'RCB'],
    'Special Teams': ['K', 'P', 'LS']
}

# Function to assign consolidated categories
def assign_consolidated_category(pos):
    for category, positions in consolidated_categories.items():
        if pos in positions:
            return category
    return 'Unknown'

# Assume 'team_results' contains pre-processed data for each team

# Add 'Team' column to each team's DataFrame in 'team_results'
for team_name, df in team_results.items():
    df['Team'] = team_name  # Assign the team name to each DataFrame

# Combine all processed data into a single DataFrame
all_data = pd.concat(team_results.values(), ignore_index=True)

# Assign consolidated categories
all_data['Consolidated_Category'] = all_data['Pos'].apply(assign_consolidated_category)

# Remove 'Unknown' categories (if any)
all_data = all_data[all_data['Consolidated_Category'] != 'Unknown']

# Aggregate data to get average Age and Weight for each consolidated category within each team
consolidated_averages = all_data.groupby(['Consolidated_Category', 'Team']).agg(
    Average_Age=('Age', 'mean'),
    Average_Weight=('Wt', 'mean')
).reset_index()

# Normalize the averages by consolidated category
consolidated_averages['Normalized_Age'] = consolidated_averages.groupby('Consolidated_Category')['Average_Age'].transform(
    lambda x: (x - x.mean()) / x.std())
consolidated_averages['Normalized_Weight'] = consolidated_averages.groupby('Consolidated_Category')['Average_Weight'].transform(
    lambda x: (x - x.mean()) / x.std())

# Rank the teams for each consolidated category by actual values (age and weight)
consolidated_averages['Rank_Age'] = consolidated_averages.groupby('Consolidated_Category')['Average_Age'].rank(ascending=False)
consolidated_averages['Rank_Weight'] = consolidated_averages.groupby('Consolidated_Category')['Average_Weight'].rank(ascending=False)

# Function to calculate combined ranks for offense, defense, and special teams
def calculate_combined_ranks(data):
    combined_ranks_list = []

    for team in data['Team'].unique():
        # Filter data by team
        team_data = data[data['Team'] == team]

        # Calculate combined scores for each category
        offense_data = team_data[team_data['Consolidated_Category'].isin(['Offensive Line', 'Quarterbacks', 'Running Backs', 'Tight Ends', 'Receivers'])]
        defense_data = team_data[team_data['Consolidated_Category'].isin(['Defensive Line', 'Linebackers', 'Defensive Backs'])]
        special_teams_data = team_data[team_data['Consolidated_Category'] == 'Special Teams']

        offense_rank_age = offense_data['Rank_Age'].sum()
        offense_rank_weight = offense_data['Rank_Weight'].sum()
        defense_rank_age = defense_data['Rank_Age'].sum()
        defense_rank_weight = defense_data['Rank_Weight'].sum()
        special_teams_rank_age = special_teams_data['Rank_Age'].sum()
        special_teams_rank_weight = special_teams_data['Rank_Weight'].sum()

        # Calculate averages and normalized values
        offense_avg_age = offense_data['Average_Age'].mean()
        offense_norm_age = offense_data['Normalized_Age'].mean()
        offense_avg_weight = offense_data['Average_Weight'].mean()
        offense_norm_weight = offense_data['Normalized_Weight'].mean()

        defense_avg_age = defense_data['Average_Age'].mean()
        defense_norm_age = defense_data['Normalized_Age'].mean()
        defense_avg_weight = defense_data['Average_Weight'].mean()
        defense_norm_weight = defense_data['Normalized_Weight'].mean()

        special_teams_avg_age = special_teams_data['Average_Age'].mean()
        special_teams_norm_age = special_teams_data['Normalized_Age'].mean()
        special_teams_avg_weight = special_teams_data['Average_Weight'].mean()
        special_teams_norm_weight = special_teams_data['Normalized_Weight'].mean()

        # Add the results to the list
        combined_ranks_list.append({
            'Team': team,
            'Offense_Rank_Age': offense_rank_age,
            'Offense_Rank_Weight': offense_rank_weight,
            'Defense_Rank_Age': defense_rank_age,
            'Defense_Rank_Weight': defense_rank_weight,
            'Special_Teams_Rank_Age': special_teams_rank_age,
            'Special_Teams_Rank_Weight': special_teams_rank_weight,
            'Offense_Avg_Age': offense_avg_age,
            'Offense_Norm_Age': offense_norm_age,
            'Offense_Avg_Weight': offense_avg_weight,
            'Offense_Norm_Weight': offense_norm_weight,
            'Defense_Avg_Age': defense_avg_age,
            'Defense_Norm_Age': defense_norm_age,
            'Defense_Avg_Weight': defense_avg_weight,
            'Defense_Norm_Weight': defense_norm_weight,
            'Special_Teams_Avg_Age': special_teams_avg_age,
            'Special_Teams_Norm_Age': special_teams_norm_age,
            'Special_Teams_Avg_Weight': special_teams_avg_weight,
            'Special_Teams_Norm_Weight': special_teams_norm_weight
        })

    # Create a DataFrame from the list
    combined_ranks = pd.DataFrame(combined_ranks_list)

    # Sort by combined ranks (lowest scores first)
    combined_ranks.sort_values(by=['Offense_Rank_Age', 'Defense_Rank_Age', 'Special_Teams_Rank_Age'], inplace=True)
    return combined_ranks

# Calculate combined ranks
combined_ranks = calculate_combined_ranks(consolidated_averages)

# Display the combined ranks in a formatted table
print("\n=== Combined Ranks for Teams ===")
print(tabulate(combined_ranks, headers='keys', tablefmt='fancy_grid'))



=== Combined Ranks for Teams ===
╒════╤══════════╤════════════════════╤═══════════════════════╤════════════════════╤═══════════════════════╤══════════════════════════╤═════════════════════════════╤═══════════════════╤════════════════════╤══════════════════════╤═══════════════════════╤═══════════════════╤════════════════════╤══════════════════════╤═══════════════════════╤═════════════════════════╤══════════════════════════╤════════════════════════════╤═════════════════════════════╕
│    │ Team     │   Offense_Rank_Age │   Offense_Rank_Weight │   Defense_Rank_Age │   Defense_Rank_Weight │   Special_Teams_Rank_Age │   Special_Teams_Rank_Weight │   Offense_Avg_Age │   Offense_Norm_Age │   Offense_Avg_Weight │   Offense_Norm_Weight │   Defense_Avg_Age │   Defense_Norm_Age │   Defense_Avg_Weight │   Defense_Norm_Weight │   Special_Teams_Avg_Age │   Special_Teams_Norm_Age │   Special_Teams_Avg_Weight │   Special_Teams_Norm_Weight │
╞════╪══════════╪════════════════════╪══════════════════════

In [529]:
import pandas as pd
from tabulate import tabulate

# Assume the consolidated_averages DataFrame is already available
# containing columns: Team, Consolidated_Category, Average_Age, Average_Weight, Normalized_Age, Normalized_Weight

# Function to create and display six separate tables
def display_corrected_tables(data):
    # Filter and rank for Offense Age
    offense_age = data[data['Consolidated_Category'].isin(['Offensive Line', 'Quarterbacks', 'Running Backs', 'Tight Ends', 'Receivers'])]
    offense_age = offense_age.groupby('Team').agg(
        Offense_Avg_Age=('Average_Age', 'mean'),
        Offense_Norm_Age=('Normalized_Age', 'mean')
    ).reset_index()
    offense_age['Offense_Rank_Age'] = offense_age['Offense_Avg_Age'].rank(ascending=False)
    offense_age.sort_values(by='Offense_Rank_Age', inplace=True)
    offense_age['Rank'] = range(1, len(offense_age) + 1)
    
    # Display Offense Age Rankings
    print("\n=== Offense Age Rankings ===")
    print(tabulate(offense_age, headers='keys', tablefmt='fancy_grid'))

    # Filter and rank for Offense Weight
    offense_weight = data[data['Consolidated_Category'].isin(['Offensive Line', 'Quarterbacks', 'Running Backs', 'Tight Ends', 'Receivers'])]
    offense_weight = offense_weight.groupby('Team').agg(
        Offense_Avg_Weight=('Average_Weight', 'mean'),
        Offense_Norm_Weight=('Normalized_Weight', 'mean')
    ).reset_index()
    offense_weight['Offense_Rank_Weight'] = offense_weight['Offense_Avg_Weight'].rank(ascending=False)
    offense_weight.sort_values(by='Offense_Rank_Weight', inplace=True)
    offense_weight['Rank'] = range(1, len(offense_weight) + 1)
    
    # Display Offense Weight Rankings
    print("\n=== Offense Weight Rankings ===")
    print(tabulate(offense_weight, headers='keys', tablefmt='fancy_grid'))

    # Filter and rank for Defense Age
    defense_age = data[data['Consolidated_Category'].isin(['Defensive Line', 'Linebackers', 'Defensive Backs'])]
    defense_age = defense_age.groupby('Team').agg(
        Defense_Avg_Age=('Average_Age', 'mean'),
        Defense_Norm_Age=('Normalized_Age', 'mean')
    ).reset_index()
    defense_age['Defense_Rank_Age'] = defense_age['Defense_Avg_Age'].rank(ascending=False)
    defense_age.sort_values(by='Defense_Rank_Age', inplace=True)
    defense_age['Rank'] = range(1, len(defense_age) + 1)
    
    # Display Defense Age Rankings
    print("\n=== Defense Age Rankings ===")
    print(tabulate(defense_age, headers='keys', tablefmt='fancy_grid'))

    # Filter and rank for Defense Weight
    defense_weight = data[data['Consolidated_Category'].isin(['Defensive Line', 'Linebackers', 'Defensive Backs'])]
    defense_weight = defense_weight.groupby('Team').agg(
        Defense_Avg_Weight=('Average_Weight', 'mean'),
        Defense_Norm_Weight=('Normalized_Weight', 'mean')
    ).reset_index()
    defense_weight['Defense_Rank_Weight'] = defense_weight['Defense_Avg_Weight'].rank(ascending=False)
    defense_weight.sort_values(by='Defense_Rank_Weight', inplace=True)
    defense_weight['Rank'] = range(1, len(defense_weight) + 1)
    
    # Display Defense Weight Rankings
    print("\n=== Defense Weight Rankings ===")
    print(tabulate(defense_weight, headers='keys', tablefmt='fancy_grid'))

    # Filter and rank for Special Teams Age
    special_teams_age = data[data['Consolidated_Category'] == 'Special Teams']
    special_teams_age = special_teams_age.groupby('Team').agg(
        Special_Teams_Avg_Age=('Average_Age', 'mean'),
        Special_Teams_Norm_Age=('Normalized_Age', 'mean')
    ).reset_index()
    special_teams_age['Special_Teams_Rank_Age'] = special_teams_age['Special_Teams_Avg_Age'].rank(ascending=False)
    special_teams_age.sort_values(by='Special_Teams_Rank_Age', inplace=True)
    special_teams_age['Rank'] = range(1, len(special_teams_age) + 1)
    
    # Display Special Teams Age Rankings
    print("\n=== Special Teams Age Rankings ===")
    print(tabulate(special_teams_age, headers='keys', tablefmt='fancy_grid'))

    # Filter and rank for Special Teams Weight
    special_teams_weight = data[data['Consolidated_Category'] == 'Special Teams']
    special_teams_weight = special_teams_weight.groupby('Team').agg(
        Special_Teams_Avg_Weight=('Average_Weight', 'mean'),
        Special_Teams_Norm_Weight=('Normalized_Weight', 'mean')
    ).reset_index()
    special_teams_weight['Special_Teams_Rank_Weight'] = special_teams_weight['Special_Teams_Avg_Weight'].rank(ascending=False)
    special_teams_weight.sort_values(by='Special_Teams_Rank_Weight', inplace=True)
    special_teams_weight['Rank'] = range(1, len(special_teams_weight) + 1)
    
    # Display Special Teams Weight Rankings
    print("\n=== Special Teams Weight Rankings ===")
    print(tabulate(special_teams_weight, headers='keys', tablefmt='fancy_grid'))

# Display the six corrected tables
display_corrected_tables(consolidated_averages)



=== Offense Age Rankings ===
╒════╤══════════╤═══════════════════╤════════════════════╤════════════════════╤════════╕
│    │ Team     │   Offense_Avg_Age │   Offense_Norm_Age │   Offense_Rank_Age │   Rank │
╞════╪══════════╪═══════════════════╪════════════════════╪════════════════════╪════════╡
│  3 │ Patriots │           26.8572 │           0.498187 │                  1 │      1 │
├────┼──────────┼───────────────────┼────────────────────┼────────────────────┼────────┤
│  2 │ Jets     │           26.8417 │          -0.295392 │                  2 │      2 │
├────┼──────────┼───────────────────┼────────────────────┼────────────────────┼────────┤
│  1 │ Dolphins │           26.7192 │           0.316268 │                  3 │      3 │
├────┼──────────┼───────────────────┼────────────────────┼────────────────────┼────────┤
│  0 │ Bills    │           26.3417 │          -0.519064 │                  4 │      4 │
╘════╧══════════╧═══════════════════╧════════════════════╧════════════════════╧═

In [530]:
import pandas as pd

# Define the path to the directory
directory_path = '/users/steventuschman/Desktop/NFL_ANALYSIS/'

# Load the CSV files while skipping the incorrect header row
offensive_stats = pd.read_csv(f'{directory_path}Offensive_Stats.csv', skiprows=1)
defensive_stats = pd.read_csv(f'{directory_path}Defensive_Stats.csv', skiprows=1)
returns_stats = pd.read_csv(f'{directory_path}Returns.csv', skiprows=1)

# Display the header and first 34 rows of each file
print("Offensive Stats - Header and First 34 Rows:")
print(offensive_stats.head(34))
print("\nDefensive Stats - Header and First 34 Rows:")
print(defensive_stats.head(34))
print("\nReturns Stats - Header and First 34 Rows:")
print(returns_stats.head(34))


Offensive Stats - Header and First 34 Rows:
   Unnamed: 0             Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4  \
0          Rk                     Tm          G         PF        Yds   
1           1         Dallas Cowboys         17        509       6317   
2           2         Miami Dolphins         17        496       6822   
3           3    San Francisco 49ers         17        491       6773   
4           4       Baltimore Ravens         17        483       6296   
5           5          Detroit Lions         17        461       6712   
6           6          Buffalo Bills         17        451       6366   
7           7    Philadelphia Eagles         17        433       6024   
8           8       Los Angeles Rams         17        404       6108   
9           9     New Orleans Saints         17        402       5732   
10         10     Indianapolis Colts         17        396       5725   
11         11       Cleveland Browns         17        396       5710   
12     

In [531]:
import pandas as pd

# Load the datasets, skipping the first row and manually setting the header from the second row
offense_data = pd.read_csv('/users/steventuschman/Desktop/NFL_ANALYSIS/Offensive_Stats.csv', header=1)
defense_data = pd.read_csv('/users/steventuschman/Desktop/NFL_ANALYSIS/Defensive_Stats.csv', header=1)
returns_data = pd.read_csv('/users/steventuschman/Desktop/NFL_ANALYSIS/Returns.csv', header=1)

# Update the column names manually from the second row (header row)
offense_data.columns = offense_data.iloc[0]  # Set the second row as the header
defense_data.columns = defense_data.iloc[0]  # Set the second row as the header
returns_data.columns = returns_data.iloc[0]  # Set the second row as the header

# Drop the duplicate header rows from the data
offense_data = offense_data[1:]
defense_data = defense_data[1:]
returns_data = returns_data[1:]

# Print the column names again to confirm
print("Offensive Stats Columns:", offense_data.columns.tolist())
print("Defensive Stats Columns:", defense_data.columns.tolist())
print("Returns Stats Columns:", returns_data.columns.tolist())


Offensive Stats Columns: ['Rk', 'Tm', 'G', 'PF', 'Yds', 'Ply', 'Y/P', 'TO', 'FL', '1stD', 'Cmp', 'Att', 'Yds', 'TD', 'Int', 'NY/A', '1stD', 'Att', 'Yds', 'TD', 'Y/A', '1stD', 'Pen', 'Yds', '1stPy', 'Sc%', 'TO%', 'EXP']
Defensive Stats Columns: ['Rk', 'Tm', 'G', 'PA', 'Yds', 'Ply', 'Y/P', 'TO', 'FL', '1stD', 'Cmp', 'Att', 'Yds', 'TD', 'Int', 'NY/A', '1stD', 'Att', 'Yds', 'TD', 'Y/A', '1stD', 'Pen', 'Yds', '1stPy', 'Sc%', 'TO%', 'EXP']
Returns Stats Columns: ['Rk', 'Tm', 'G', 'Ret', 'Yds', 'TD', 'Lng', 'Y/R', 'Rt', 'Yds', 'TD', 'Lng', 'Y/Rt', 'APYd']


In [532]:
# Define the teams of interest
teams = ['Buffalo Bills', 'Miami Dolphins', 'New York Jets', 'New England Patriots']

# Extract relevant columns for each dataset
offense_stats = offense_data[offense_data['Tm'].isin(teams)][['Tm', 'PF']]
defense_stats = defense_data[defense_data['Tm'].isin(teams)][['Tm', 'PA']]
returns_stats = returns_data[returns_data['Tm'].isin(teams)][['Tm', 'Y/Rt']]

# Display the extracted data
print("Offense Stats for Selected Teams:")
print(offense_stats)

print("\nDefense Stats for Selected Teams:")
print(defense_stats)

print("\nSpecial Teams Stats for Selected Teams:")
print(returns_stats)


Offense Stats for Selected Teams:
0                     Tm   PF
2         Miami Dolphins  496
6          Buffalo Bills  451
29         New York Jets  268
31  New England Patriots  236

Defense Stats for Selected Teams:
0                     Tm   PA
4          Buffalo Bills  311
12         New York Jets  355
15  New England Patriots  366
22        Miami Dolphins  391

Special Teams Stats for Selected Teams:
0                     Tm  Y/Rt
10         New York Jets  22.6
11  New England Patriots  26.2
15         Buffalo Bills  20.5
24        Miami Dolphins  24.5


In [533]:
import pandas as pd
from scipy.stats import pearsonr

# Data for correlation analysis
offense_data = {
    'Team': ['Buffalo Bills', 'Miami Dolphins', 'New York Jets', 'New England Patriots'],
    'Average_Age_Offense': [26.34, 26.72, 26.84, 26.86],  # Use actual data values
    'Average_Weight_Offense': [240.22, 236.54, 240.69, 239.38],  # Use actual data values
    'PF': [451, 496, 268, 236]  # Points For
}

defense_data = {
    'Team': ['Buffalo Bills', 'Miami Dolphins', 'New York Jets', 'New England Patriots'],
    'Average_Age_Defense': [27.90, 27.81, 26.54, 26.78],  # Use actual data values
    'Average_Weight_Defense': [239.84, 245.77, 237.63, 246.96],  # Use actual data values
    'PA': [311, 391, 355, 366]  # Points Allowed
}

special_teams_data = {
    'Team': ['Buffalo Bills', 'Miami Dolphins', 'New York Jets', 'New England Patriots'],
    'Average_Age_ST': [29.33, 26.67, 35.00, 26.33],  # Use actual data values
    'Average_Weight_ST': [209.67, 212.33, 222.00, 218.33],  # Use actual data values
    'Y/Rt': [20.5, 24.5, 22.6, 26.2]  # Yards per Return
}

# Convert to DataFrames
offense_df = pd.DataFrame(offense_data)
defense_df = pd.DataFrame(defense_data)
special_teams_df = pd.DataFrame(special_teams_data)

# Calculate Pearson correlations
def calculate_pearson(df, x, y):
    corr, p_value = pearsonr(df[x], df[y])
    return corr, p_value

# Offense correlations
offense_age_pf_corr, offense_age_pf_pval = calculate_pearson(offense_df, 'Average_Age_Offense', 'PF')
offense_weight_pf_corr, offense_weight_pf_pval = calculate_pearson(offense_df, 'Average_Weight_Offense', 'PF')

# Defense correlations
defense_age_pa_corr, defense_age_pa_pval = calculate_pearson(defense_df, 'Average_Age_Defense', 'PA')
defense_weight_pa_corr, defense_weight_pa_pval = calculate_pearson(defense_df, 'Average_Weight_Defense', 'PA')

# Special Teams correlations
st_age_yrt_corr, st_age_yrt_pval = calculate_pearson(special_teams_df, 'Average_Age_ST', 'Y/Rt')
st_weight_yrt_corr, st_weight_yrt_pval = calculate_pearson(special_teams_df, 'Average_Weight_ST', 'Y/Rt')

# Display results
print("Offense - Age vs PF Correlation:", offense_age_pf_corr, "P-Value:", offense_age_pf_pval)
print("Offense - Weight vs PF Correlation:", offense_weight_pf_corr, "P-Value:", offense_weight_pf_pval)
print("Defense - Age vs PA Correlation:", defense_age_pa_corr, "P-Value:", defense_age_pa_pval)
print("Defense - Weight vs PA Correlation:", defense_weight_pa_corr, "P-Value:", defense_weight_pa_pval)
print("Special Teams - Age vs Y/Rt Correlation:", st_age_yrt_corr, "P-Value:", st_age_yrt_pval)
print("Special Teams - Weight vs Y/Rt Correlation:", st_weight_yrt_corr, "P-Value:", st_weight_yrt_pval)


Offense - Age vs PF Correlation: -0.6662419458096291 P-Value: 0.33375805419037086
Offense - Weight vs PF Correlation: -0.5915624896794154 P-Value: 0.4084375103205846
Defense - Age vs PA Correlation: -0.19484890284182624 P-Value: 0.8051510971581737
Defense - Weight vs PA Correlation: 0.6255686110546927 P-Value: 0.37443138894530725
Special Teams - Age vs Y/Rt Correlation: -0.5370547938067909 P-Value: 0.4629452061932091
Special Teams - Weight vs Y/Rt Correlation: 0.3905370333714704 P-Value: 0.6094629666285296


#### **Conclusion**

The objective of this analysis was to explore potential correlations between player attributes (such as average age and weight across different positions) and key team performance metrics from the 2023 NFL season. Specifically, the study focused on four AFC East teams: the Buffalo Bills, Miami Dolphins, New York Jets, and New England Patriots. The relationships investigated included:

Offensive Attributes: Correlation between the average age and weight of offensive players and the total points scored by the team (Points For, PF).
Defensive Attributes: Correlation between the average age and weight of defensive players and the total points allowed by the team (Points Against, PA).
Special Teams Attributes: Correlation between the average age and weight of special teams players and the average yards per return (Y/Rt).
Results and Interpretation
The results of the Pearson correlation analysis were as follows:

Offense - Age vs Points For (PF):

Correlation: -0.67
P-Value: 0.33
Interpretation: There is a moderate negative correlation between the average age of offensive players and points scored (PF). However, the p-value of 0.33 suggests that this correlation is not statistically significant, as there is a 33% chance that this result could occur by random chance.
Offense - Weight vs Points For (PF):

Correlation: -0.59
P-Value: 0.41
Interpretation: A moderate negative correlation was observed between the average weight of offensive players and points scored (PF). The p-value of 0.41 is relatively high, indicating that there is a 41% likelihood that this correlation occurred by random chance, thus rendering it statistically insignificant.
Defense - Age vs Points Against (PA):

Correlation: -0.19
P-Value: 0.81
Interpretation: The correlation between the average age of defensive players and points allowed (PA) is weakly negative, with a correlation coefficient of -0.19. The p-value of 0.81 is very high, meaning there is an 81% probability that this correlation could have occurred by random chance, further confirming its insignificance.
Defense - Weight vs Points Against (PA):

Correlation: 0.63
P-Value: 0.37
Interpretation: There is a moderate positive correlation between the average weight of defensive players and points allowed (PA). However, the p-value of 0.37 suggests that there is a 37% chance that this observed correlation could occur by random chance, indicating that it is not statistically significant.
Special Teams - Age vs Yards per Return (Y/Rt):

Correlation: -0.54
P-Value: 0.46
Interpretation: The correlation between the average age of special teams players and yards per return (Y/Rt) is moderately negative. The p-value of 0.46 indicates a 46% chance that this correlation could have occurred by random chance, making it statistically insignificant.
Special Teams - Weight vs Yards per Return (Y/Rt):

Correlation: 0.39
P-Value: 0.61
Interpretation: A weak positive correlation was found between the average weight of special teams players and yards per return (Y/Rt). With a p-value of 0.61, there is a 61% probability that this result could have occurred by random chance, confirming that this relationship is not statistically significant.
Conclusion
Overall, the findings did not reveal any statistically significant correlations between the analyzed player attributes and team performance metrics. The moderately strong correlation coefficients observed in some cases (such as offense age vs. Points For and defense weight vs. Points Against) were negated by high p-values, indicating a substantial probability that these results could occur due to random chance. Therefore, there is insufficient evidence to suggest a meaningful relationship between these metrics within the examined teams.

Given the lack of significant findings, the project concludes without expanding to additional teams, as further analysis would not likely yield more meaningful results. The initial hypothesis that no significant correlations would be present in the data appears to be validated by these results.







